chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -1624,3 +1652,149 @@ pub(crate) fn parse_block_triggers(
     }
     triggers
 }
+
+async fn fetch_receipt_from_ethereum_client(
+    eth: &EthereumAdapter,
+    transaction_hash: &H256,
+) -> anyhow::Result<TransactionReceipt> {
+    match eth
+        .web3
+        .eth()
+        .transaction_receipt(*transaction_hash)
+        .compat()
+        .await
+    {
+        Ok(Some(receipt)) => Ok(receipt),
+        Ok(None) => bail!("Could not find transaction receipt"),
+        Err(error) => bail!("Failed to fetch transaction receipt: {}", error),
+    }
+}
+
+async fn filter_call_triggers_from_unsuccessful_transactions(
+    mut block: BlockWithTriggers<crate::Chain>,
+    eth: &EthereumAdapter,
+    chain_store: &Arc<dyn ChainStore>,
+) -> anyhow::Result<BlockWithTriggers<crate::Chain>> {
+    // First, we separate call triggers from other trigger types
+    let calls: Vec<&Arc<EthereumCall>> = block
+        .trigger_data
+        .iter()
+        .filter_map(|trigger| match trigger {
+            EthereumTrigger::Call(call_trigger) => Some(call_trigger),
+            _ => None,
+        })
+        .collect();
+
+    // Then we get the transaction hash for each call
+    let transaction_hashes = calls
+        .iter()
+        .map(|call| call.transaction_hash)
+        .collect::<Option<BTreeSet<H256>>>()
+        .ok_or(anyhow!(
+            "failed to obtain transaction hash from call triggers"
+        ))?;
```

Comment:
It's not clear to me when an `EthereumCall` might have a `None` for its `transaction_hash` field.

Since that information is required for the rest of the function, I'm considering its absence to be an error.